libs/muddle/src/kademlia/peer_tracker.cpp

```diff
@@ -407,41 +418,54 @@ void PeerTracker::PullPeerKnowledge()
     }
   }
 
+  // Searching in parallel to different nodes
+  int64_t queue_size    = 0;
+  int64_t promise_count = 0;
+  {
+    FETCH_LOCK(pull_mutex_);
+    queue_size    = static_cast<int64_t>(peer_pull_queue_.size());
+    promise_count = static_cast<int64_t>(pull_promises_.size());
+  }
+
   // If it is still empty
-  if (peer_pull_queue_.empty())
+  if (queue_size == 0)
   {
     return;
   }
 
-  // Searching in parallel to different nodes
-  int64_t tasks_to_setup =
-      tracker_configuration_.max_discovery_tasks - static_cast<int64_t>(pull_promises_.size());
-
-  tasks_to_setup = std::min(tasks_to_setup, static_cast<int64_t>(peer_pull_queue_.size()));
+  int64_t tasks_to_setup = cfg.max_discovery_tasks - promise_count;
+  tasks_to_setup         = std::min(tasks_to_setup, queue_size);
 
   for (int64_t task = 0; task < tasks_to_setup; ++task)
   {
-    auto address = peer_pull_queue_.front();
-    if (address.size() != Packet::ADDRESS_SIZE)
+    Address search_for{};
+    Address address{};
+
     {
-      continue;
-    }
+      FETCH_LOCK(pull_mutex_);
+      address = peer_pull_queue_.front();
 
-    auto search_for = peer_pull_map_[address];
+      if (address.size() != Packet::ADDRESS_SIZE)
+      {
+        continue;
```

Comment:
Is it a dead loop? If address is invalid, it is not popped, so considered over and over again.

Reply:
Thanks. That is a really good observation.